Change build rules to stop mobile and fixed-line descs inheriting from the generalDesc element

lararennie opened the pull request:
Metadata changes:
-- Includes the fixed-line IR change that was submitted on open-source but that we hadn't updated the files for until this point of time.
-- Drops the "-1" that was erroneously included in possible lengths before (didn't break anything, but was wrong) - it was a possibleLength of a sub-component so got added to the generalDesc possibleLengths
-- possibleNumberPattern no longer inherited: we don't use this anyway, we will do another CL soon to stop including it at all in the generated metadata
-- exampleNumber is no longer set on fixed-line and mobile elements from the generalDesc

XML file changes:
-- Stopped specifying "NA" and "-1" specifically for fixed-line and mobile blocks; now they are treated as every other type of phone number: if missing, don't fill them in from generalDesc, but leave them missing.

Code changes:
-- Stop using the exampleNumber on generalDesc for non-geo entities, but look at their phonenumber descs - the exampleNumber won't be stored on the generalDesc anymore. This affects porters if they either copied our build logic or used our built metadata in some way; they should update this method in their port too.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlei18n/libphonenumber/1422)
<!-- Reviewable:end -->

lararennie commented:
Have fun!

andyst commented:
@Iran metdata change:
Please remove that from pending_code_changes.txt.
 - That is quite out of place in this pull-request.
 - It will conflict with the metadata release script in terms of auto-generated release notes. I guess this refers to "Removing fixed-line prefixes 57 and 64 for North Khorasan". We'll mention IR automatically.

lararennie commented:
You won't actually, that is why it's here. You won't because no IR file
will be generated automatically. We have had this before and a manual merge
is the only option.

7. nov. 2016 6.14 p.m. skrev "Andy Staudacher" <notifications@github.com>:

> @Iran <https://github.com/Iran> metdata change:
> Please remove that from pending_code_changes.txt.
>
>    - That is quite out of place in this pull-request.
>    - It will conflict with the metadata release script in terms of
>    auto-generated release notes. I guess this refers to "Removing fixed-line
>    prefixes 57 and 64 for North Khorasan". We'll mention IR automatically.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlei18n/libphonenumber/pull/1422#issuecomment-258898812>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIjis7AbpiHbbHToN3fmeD4S2giWqmKgks5q71yMgaJpZM4KrPLh>
> .
>

lararennie commented:
Only option is a separate PR, but pending code changes will still have to
be edited.

7. nov. 2016 6.28 p.m. skrev "Lara Rennie" <lararennie@google.com>:

> You won't actually, that is why it's here. You won't because no IR file
> will be generated automatically. We have had this before and a manual merge
> is the only option.
>
> 7. nov. 2016 6.14 p.m. skrev "Andy Staudacher" <notifications@github.com>:
>
>> @Iran <https://github.com/Iran> metdata change:
>> Please remove that from pending_code_changes.txt.
>>
>>    - That is quite out of place in this pull-request.
>>    - It will conflict with the metadata release script in terms of
>>    auto-generated release notes. I guess this refers to "Removing fixed-line
>>    prefixes 57 and 64 for North Khorasan". We'll mention IR automatically.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/googlei18n/libphonenumber/pull/1422#issuecomment-258898812>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AIjis7AbpiHbbHToN3fmeD4S2giWqmKgks5q71yMgaJpZM4KrPLh>
>> .
>>
>

keghani commented on java/pending_code_changes.txt:
```diff
@@ -1,6 +1,21 @@
+   removed.
+   There should no longer be an exampleNumber at the generalDesc level, but it
+   should be present at every PhoneNumberDesc with data.
+Metadata changes
```
I don't see a commit in master with an IR change since the last release? https://github.com/googlei18n/libphonenumber/commits/master

If it's a pending pull request let's not merge until this PR is in. I say this PR is priority, and we merge it before any other open PRs at this point; since it's be a pain to sync with the base branch this otherwise.

In case I'm missing something and there is an IR change:
We'll have to merge this manually anyway, but it's less likely to be missed or have formatting errors by the merger if it's last (after code changes) since that's where the rest of the metadata changes will show.

And since you're using the proper format anyway, optionally add a ':' after changes :)

lararennie commented:
Wow:

https://github.com/googlei18n/libphonenumber/pull/1418

So thought I'd merged it ;)

If not then yes indeed can do after this ;)

It's been submitted internally so I want to do it before the cut to give
him the credit....!!

- I'll make the changes
- no idea re nanoproto.... Can remove the jars.

8. nov. 2016 11.31 a.m. skrev "Keghani Kouzoujian" <notifications@github.com
>:

> *@keghani* requested changes on this pull request.
>
> (1) Please also add generated java/libphonenumber/src/com/
> google/i18n/phonenumbers/data/ShortNumberMetadataProto_* files.
>
> (2) No action, but FYI:
> I get different .class files in cpp-build jar, although the resulting C++
> metadata files are the same as in this PR (and pkgdiff doesn't pick up the
> .class file diffs): ~keghani/GitHub_diffs/keghani_
> output_for_%231422/cppdiff
> ~keghani/GitHub_diffs/keghani_output_for_%231422/pkgdiff_
> reports/cpp-build/1.0-SNAPSHOT-jar-with-dependencies_to_1.0-SNAPSHOT-
> jar-with-dependencies/changes_report.html#InformationFiles
>
> (3) You're checking in nano files in the java-build jar, see how mine's
> different:
> ~keghani/GitHub_diffs/keghani_output_for_%231422/pkgdiff_
> reports/java-build/1.0-SNAPSHOT-jar-with-dependencies_to_1.0-SNAPSHOT-
> jar-with-dependencies/changes_report.html
> Here's the raw diff (not pkgdiff) to see that the .class files are
> different (ignore the metadata files, I just ran mvn package at the end
> with the generated metadata files which you didn't check in):
> ~keghani/GitHub_diffs/keghani_output_for_%231422/javadiff
>
> I think there's something about your cpp setup that we should fix, or
> you're reusing an old clone.
> For now we can go with this PR and I can send another one right afterwards
> with updated jars - we definitely don't want nano in there (I don't know if
> the Maven release jars would pick that up from the build jars).
> The important thing to ensure this PR has correct changes (IMO) is tests
> pass, everything gets generated the same between both our setups even with
> different build jars; just that the jars are different (and we don't
> release these jars), which I can hack by following up with a new change for
> now.
> ------------------------------
>
> In tools/java/common/test/com/google/i18n/phonenumbers/
> BuildMetadataFromXmlTest.java
> <https://github.com/googlei18n/libphonenumber/pull/1422#pullrequestreview-7586566>
> :
>
> > @@ -612,8 +591,7 @@ public void testFilterMetadata_fullBuild() throws Exception {
>      assertTrue(metadataCollection.getMetadataCount() == 1);
>      PhoneMetadata metadata = metadataCollection.getMetadataList().get(0);
>      assertTrue(metadata.hasGeneralDesc());
> -    assertTrue(metadata.getGeneralDesc().hasExampleNumber());
> -    assertEquals("10123456", metadata.getGeneralDesc().getExampleNumber());
> +    assertFalse(metadata.getGeneralDesc().hasExampleNumber());
>
> Please add line after this for assertEquals("", metadata.getGeneralDesc().
> getExampleNumber());
> ------------------------------
>
> In java/pending_code_changes.txt
> <https://github.com/googlei18n/libphonenumber/pull/1422#pullrequestreview-7586566>
> :
>
> > @@ -1,6 +1,18 @@
>  GitHub project changes:
>   - Changed tag to vX.Y.Z from libphonenumber-X.Y.Z; this may affect ports and
>     derived projects.
> +Metadata structure changes:
>
> I would be explicit about whether we're talking about the XML or proto or
> both for each restriction, since they don't have the same rules.
> ------------------------------
>
> In java/pending_code_changes.txt
> <https://github.com/googlei18n/libphonenumber/pull/1422#pullrequestreview-7586566>
> :
>
> > @@ -1,6 +1,21 @@
>  GitHub project changes:
>   - Changed tag to vX.Y.Z from libphonenumber-X.Y.Z; this may affect ports and
>     derived projects.
> +Metadata structure changes:
> + - Mobile and Fixed-Line blocks are no longer assumed to inherit missing data
> +   from the GeneralDesc, but are treated like every other phone number type.
> +   This means that for the non-geographical country codes, like +800, the
> +   example number has been moved from generalDesc to the relevant number types,
> +   and the code in getExampleNumberForNonGeoEntity has been changed to look at
> +   these sub-types for an example number.
> +   This also means that the "NA" and "-1" blocks present in the metadata to
> +   indicate that no mobile or fixed-line numbers appear for the entity have been
> +   removed.
> +   There should no longer be an exampleNumber at the generalDesc level, but it
> +   should be present at every PhoneNumberDesc with data.
> +Metadata changes
>
> I don't see a commit in master with an IR change since the last release?
> https://github.com/googlei18n/libphonenumber/commits/master
>
> If it's a pending pull request let's not merge until this PR is in. I say
> this PR is priority, and we merge it before any other open PRs at this
> point; since it's be a pain to sync with the base branch this otherwise.
>
> In case I'm missing something and there is an IR change:
> We'll have to merge this manually anyway, but it's less likely to be
> missed or have formatting errors by the merger if it's last (after code
> changes) since that's where the rest of the metadata changes will show.
>
> And since you're using the proper format anyway, optionally add a ':'
> after changes :)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/googlei18n/libphonenumber/pull/1422#pullrequestreview-7586566>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIjis3zyhhyL6vhrOJo_TSMP84DXeWvwks5q8E-MgaJpZM4KrPLh>
> .
>

keghani commented:
Re. jar files, please don't remove from PR! They're still used to generate the metadata files, so should be in here, it's weird to generate metadata with tools that aren't checked in :)

Re. #1418 - if you're happy with it one of us can follow up and merge after this one (before the release), in case you had to go.

keghani commented:
I thought the .class file difference may have been because I compiled with java 8, but no; with java 7 the .class files are still different between you and me.

Could be that my .class files are wrong, but your setup is still giving the nano stuff in the java-build jar, so if you don't mind I'll follow up by checking in my jars for now.

andyst commented:
Don't have anything to add. Looks good modulo Keghani's comments.
Thanks for the detailed review, Keghani!

keghani commented:
@lararennie please ping with "PTAL" or similar once you're done with new commits

keghani commented:
@lararennie I'm assuming this is ready for the next look, the commits look good.
Just running some sanity checks on my side then I'll approve and merge.

keghani commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 210 of 212 files at r1, 2 of 2 files at r3, 236 of 236 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/googlei18n/libphonenumber/1422#-:-KW3Jw6_Lbt3qHziWN7K:bnfp4nl)*
<!-- Sent from Reviewable.io -->

keghani commented:
This probably won't pass the checks as CLA is pending; I sent https://github.com/googlei18n/libphonenumber/pull/1427
